src/qforte/utils/qbt_tapering.py

```diff
@@ -0,0 +1,321 @@
+
+    ## Construct individual unitary matrices
+    untrs = []
+
+    for idx, generator in enumerate(gnrtrs):
+        circ = qf.Circuit()
+        circ.add_gate(qf.gate('X', sigma_x[idx]))
+        oprtr = qf.QubitOperator()
+        oprtr.add_term(1/np.sqrt(2), generator)
+        oprtr.add_term(1/np.sqrt(2), circ)
+        untrs.append(oprtr)
+
+    ## Construct final unitary matrix
+    unitary = qf.QubitOperator()
+    unitary.add_op(untrs[0])
+    for op in untrs[1:]:
+        unitary.operator_product(op, True, True)
+
+    # 5) Compute the various signs corresponding to the eigenvalues of the sigma_x operators that will be removed
+
+    signs = list(product([1, -1], repeat=len(sigma_x)))
+
+    return gnrtrs, sigma_x, untrs, unitary, signs
+
+def qubit_tapering_oprtr(tapered_qubits, sign, operator, unitary):
+    """
+    This function uses the unitary matrix obtained using find_Z2_symmetries to transform a given operator.
+    After the unitary transformation, the resulting operator acts on the qubits to be tapered off by at most
+    a single sigma_x operator. The sigma_x gates acting on qubits to be tapered off are replaced by their
+    +/- 1 eigenvalues, which are provided by the user.
+
+    Arguments
+
+    tapered_qubits: list of integers
+        The ID numbers of the qubits to be tapered off.
+
+    sign: list of +/- 1.
+        The desired sign structure corresponding to the eigenvalues of the sigma_x operators acting on
+        the qubits to be tapered off.
+
+    operator: QubitOperator
+        The operator that we wish to transform. For example, the qubit Hamiltonian or the UCC ansatz.
+
+    unitary: QubitOperator
+        The unitary operator that will perform the dersired transformation.
+    """
+
+    # Similarity transform the given operator using the fact that the unitary matrix is also Hermitian
+
+    ## This is necessary; otherwise later on we would be modifying the "unitary" object directly, which we do not want to do
+    unitary_tmp = qf.QubitOperator()
+    unitary_tmp.add_op(unitary)
+
+    ## This is necessary; if we do not do this, when transforming the Hamiltonian we would be changing the "hamiltonian"
+    ## attribute of the Molecule class to nonsense.
+    operator_tmp = qf.QubitOperator()
+    operator_tmp.add_op(operator)
+
+    operator_tmp.operator_product(unitary_tmp, True, True)
+    unitary_tmp.operator_product(operator_tmp, True, True)
+    operator_tmp = unitary_tmp
+
+    # Create a QForte QubitOperator object that will store the tapered operator
+    operator_tapered = qf.QubitOperator()
+
+    for i, (coeff, circuit) in enumerate(operator_tmp.terms()):
+        # Create circuit that will store the pauli string with gates corresponding to qubits to be tapered off removed
+        circ = qf.Circuit()
+        # List that will store the coefficients modified according to the sign structure provided by the user
+        for pauli in circuit.gates():
```

Comment:
Ugly again. We should define an `__iter__` for a circuit. Again, separate issue.